dot/sync/message.go

```diff
@@ -25,122 +25,154 @@ import (
+var maxResponseSize uint32 = 128 // maximum number of block datas to reply with in a BlockResponse message.
 
 // CreateBlockResponse creates a block response message from a block request message
 func (s *Service) CreateBlockResponse(blockRequest *network.BlockRequestMessage) (*network.BlockResponseMessage, error) {
-	var startHash common.Hash
-	var endHash common.Hash
+	var (
+		startHash, endHash     common.Hash
+		startHeader, endHeader *types.Header
+		err                    error
+		respSize               uint32
+	)
 
 	if blockRequest.StartingBlock == nil {
 		return nil, ErrInvalidBlockRequest
 	}
 
+	if blockRequest.Max != nil && blockRequest.Max.Exists() {
+		respSize = blockRequest.Max.Value()
+		if respSize > maxResponseSize {
+			respSize = maxResponseSize
+		}
+	} else {
+		respSize = maxResponseSize
+	}
+
 	switch startBlock := blockRequest.StartingBlock.Value().(type) {
 	case uint64:
 		if startBlock == 0 {
 			startBlock = 1
 		}
-		block, err := s.blockState.GetBlockByNumber(big.NewInt(0).SetUint64(startBlock))
+
+		block, err := s.blockState.GetBlockByNumber(big.NewInt(0).SetUint64(startBlock)) //nolint
 		if err != nil {
 			return nil, err
 		}
 
+		startHeader = block.Header
 		startHash = block.Header.Hash()
 	case common.Hash:
 		startHash = startBlock
+		startHeader, err = s.blockState.GetHeader(startHash)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	if blockRequest.EndBlockHash != nil && blockRequest.EndBlockHash.Exists() {
 		endHash = blockRequest.EndBlockHash.Value()
+		endHeader, err = s.blockState.GetHeader(endHash)
+		if err != nil {
+			return nil, err
+		}
 	} else {
-		endHash = s.blockState.BestBlockHash()
-	}
+		endNumber := big.NewInt(0).Add(startHeader.Number, big.NewInt(int64(respSize-1)))
+		bestBlockNumber, err := s.blockState.BestBlockNumber()
+		if err != nil {
+			return nil, err
+		}
 
-	startHeader, err := s.blockState.GetHeader(startHash)
-	if err != nil {
-		return nil, err
-	}
+		if endNumber.Cmp(bestBlockNumber) == 1 {
+			endNumber = bestBlockNumber
+		}
 
-	endHeader, err := s.blockState.GetHeader(endHash)
-	if err != nil {
-		return nil, err
+		endBlock, err := s.blockState.GetBlockByNumber(endNumber)
+		if err != nil {
+			return nil, err
+		}
+		endHeader = endBlock.Header
+		endHash = endHeader.Hash()
 	}
 
 	logger.Debug("handling BlockRequestMessage", "start", startHeader.Number, "end", endHeader.Number, "startHash", startHash, "endHash", endHash)
 
-	// get sub-chain of block hashes
-	subchain, err := s.blockState.SubChain(startHash, endHash)
-	if err != nil {
-		return nil, err
-	}
-
-	if len(subchain) > int(maxResponseSize) {
-		subchain = subchain[:maxResponseSize]
-	}
-
-	logger.Trace("subchain", "start", subchain[0], "end", subchain[len(subchain)-1])
-
 	responseData := []*types.BlockData{}
 
-	// TODO: check ascending vs descending direction
-	for _, hash := range subchain {
+	// ascending (ie child to parent)
+	if blockRequest.Direction == 0 {
```

Comment:
```suggestion
	switch blockRequest.Direction {
	    case 0:
	    case 1:
	}
```
nit: I like switch statements when handling finite cases.